// @juandardilag/value-objects/dist/StringValueObject/StringValueObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StringLengthOptionsDefault = exports.StringValueObject = void 0;
const ValueObject_1 = require("../ValueObject/ValueObject");
const StringValueObjectOperator_1 = require("./StringValueObjectOperator");
const StringValueObjectTyper_1 = require("./StringValueObjectTyper");
const StringValueObjectValidator_1 = require("./StringValueObjectValidator");
class StringValueObject extends ValueObject_1.ValueObject {
    constructor(_value, _operator, _typer, _validator) {
        super(_operator || new StringValueObjectOperator_1.StringValueObjectOperator(), _typer || new StringValueObjectTyper_1.StringValueObjectTyper(), _validator || new StringValueObjectValidator_1.StringValueObjectValidator(exports.StringLengthOptionsDefault), _value);
    }
    static empty() {
        return new StringValueObject("");
    }
    isEmpty() {
        return this._value == "";
    }
    is(value) {
        return this.value === value;
    }
}
exports.StringValueObject = StringValueObject;
exports.StringLengthOptionsDefault = {
    minLength: 0,
    maxLength: 100000,
};
