// @juandardilag/value-objects/dist/ValueObject/ValueObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValueObject = void 0;
const InvalidArgumentError_1 = require("./errors/InvalidArgumentError");
class ValueObject {
    constructor(_operable, _typable, _validatable, _value) {
        this._operable = _operable;
        this._typable = _typable;
        this._validatable = _validatable;
        this._value = _value;
        const validation = this.validate(_value);
        if (validation instanceof Error) {
            throw validation;
        }
    }
    get value() {
        return this._value;
    }
    set value(value) {
        if (!this.validate(value)) {
            throw new InvalidArgumentError_1.InvalidArgumentError(this.constructor.name, value);
        }
        this._value = value;
    }
    equal(other) {
        if (other === null || other === undefined) {
            return false;
        }
        if (this.constructor.name !== other.constructor.name) {
            return false;
        }
        return this.value === other.value;
    }
    is(o) {
        return this._value === o;
    }
    toString() {
        if (this._value) {
            return this._value.toString();
        }
        return this._value;
    }
    toType(val) {
        return this._typable.toType(val);
    }
    fromType(val) {
        return this._typable.fromType(val);
    }
    validate(val) {
        const validation = this._validatable.validate(val);
        if (validation instanceof Error) {
            return validation;
        }
        if (validation === false) {
            return new InvalidArgumentError_1.InvalidArgumentError(this.constructor.name, val);
        }
    }
    add(other) {
        return this._operable.add(this, other);
    }
}
exports.ValueObject = ValueObject;
